### ISSUE TITLE
Validity of integer signal processing

### ISSUE THREAD

**Issue by olofson** [olofson/audiality2#28]
## The situation

Let's face it: High quality integer/fixed point signal processing is one major pain in certain regions to implement! Any non-trivial operation becomes a nightmare of juggling bits, weighing headroom against resolution through every single step.

Not only that, but it's already been quite a few years since floating point started turning out to be faster than fixed point more often than not on desktop/workstation CPUs, and with the latest advancements, floating point SIMD instructions and all, that's probably even more true now.

However, what about handheld devices and these new low end consoles? It seems that only the latest generation of ARM CPUs have floating point hardware, and how fast are they, actually? What about Atom and other low-power x86 compatibles? When will handheld devices arrive at the point where floating point is as fast as, or faster than fixed point?

Another aspect is that Audiality 2 may not be the best option for low powered devices anyway, regardless of how the processing is done. After all, it's a modular synthesizer with sub-sample accurate scripting! A simpler, more traditional sound engine would most likely be faster, and probably easier to use as well.

Then again, if you already have an application written towards Audiality 2, porting it could be a (relatively) simple matter of converting the scripts to use more pre-rendering and/or mp3/ogg streaming, and just leaving the application code as is.
## What we need to decide

Is there any point in bothering with integer processing at all? Do any _viable_ target platforms have poor enough floating point performance that it makes sense to keep some integer support?
## A possible solution

If there is any point in doing integer processing at all, what we could do is restrict the integer support to only a small core set of voice units. This way, we have what we need to run appropriately downscaled Audiality 2 projects, without implementing integer versions of the more complex (and most likely too CPU intensive) units.

Of course, if we decide to drop integer processing altogether, then that's all there is to it. Simply don't use more and heavier units than your target platform can handle - as always!

**Comment by olofson** [comment]
Depends on #230.

**Comment by olofson** [comment]
Now that [Kobo Redux](https://github.com/olofson/koboredux) is about to be released for early access, this is becoming a rather pressing issue. It would be nice to have a demo running in browsers, but due to the 64 bit integer math here and there, A2 runs REALLY slow as asm.js...

**Comment by olofson** [comment]
It appears that WebAssembly has proper integer arithmetic, including 64 bit, which should make this less of a showstopper issue for deployment to web browsers. (However, floating point is probably the way to go, for most targets, and in the long run.)

**Comment by olofson** [comment]
There is one major performance issue we'll have to deal with: Denormals. On native builds, we can [handle this in hardware](http://softwareengineering.stackexchange.com/questions/337246/how-to-stop-denormals-from-messing-up-a-real-time-process-like-an-audio-app-or) on most/all platforms. For asm.js, we may have to resort to [traditional methods](http://www.musicdsp.org/files/denormal.pdf). Don't know about WebAssembly.

**Comment by olofson** [comment]
Long time no see... On the upside, I think that means we can pretty much forget about integer/fixed point processing. I don't see Audiality as a viable solution on any sort of hardware that would perform better with integers than with floating point, so I'll just go ahead and port to floating point. One can always revive the old code at a later date, should it ever become relevant again.
